nixos/modules/services/misc/ollama.nix

```diff
@@ -159,12 +165,16 @@ in
         OLLAMA_HOST = "${cfg.host}:${toString cfg.port}";
         HSA_OVERRIDE_GFX_VERSION = lib.mkIf (cfg.rocmOverrideGfx != null) cfg.rocmOverrideGfx;
       };
-      serviceConfig = {
+      serviceConfig = (if (cfg.user == null) then {
+        DynamicUser = true;
+      } else {
+        User = cfg.user;
+        Group = cfg.group;
+       }) // {
         ExecStart = "${lib.getExe ollamaPackage} serve";
         WorkingDirectory = cfg.home;
         StateDirectory = [ "ollama" ];
-        DynamicUser = cfg.sandbox;
-        ReadWritePaths = cfg.writablePaths;
+        ReadWritePaths = [ cfg.models ];
```

Comment:
Is there any other path that commonly needs to be writable?

Reply:
Not as far as I know, but I'm not the one who needs these (I don't even use ollama anymore). I specifically think that you should leave `writablePaths` as is, and simply have `cfg.models` as default.
``` nix
writablePaths = lib.mkOption {
  type = types.listOf types.str;
  default = [ cfg.models ];
  # omitted
};
```

Reply:
Again, the ollama module is an outlier here. https://search.nixos.org/options?channel=24.05&size=50&sort=relevance&type=packages&query=writablePaths

There is always the direct way via `systemd.services.ollama.serviceConfig.ReadWritePaths` that works for every service out there.

Reply:
> Again, the ollama module is an outlier here.

That's fine, there's nothing wrong with being an outlier. Remember, before anything becomes a convention, it begins as just a few outliers, which eventually becomes common practice.

If you have genuine reasons why the options should be changed and removed, I'm listening. But I haven't heard much other than "It's not common practice", which isn't a very logical argument.

> There is always the direct way via `systemd.services.ollama.serviceConfig.ReadWritePaths` that works for every service out there.

Yes, but not every nix user knows about that. In fact, I used nix for over a year, maybe two, before I learned that services could be overridden like that. I don't think it's especially common knowledge among non-contributors, which are the vast majority of total nix users.

When I can, I prefer to make an interface that is as simple and convenient as possible while still allowing as many possibilities as are useful.

Reply:
If the only directory that needs to be writable is the model directory, then it is not a valuable option to have and should be removed.

And looking at part 2 of [RFC42](https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md#valuable-options), the option is neither of popular, mandatory, or sensitive.

Just because we can expose anything as an option, doesn't mean we should: https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md#part-2

Reply:
> If the only directory that needs to be writable is the model directory, then it is not a valuable option to have and should be removed.

Fair if true, though that is an assumption that you're making.

@suhr could you give your opinion on this? Is the `model` directory the only place that's likely to be useful to write to through the `dynamicUser` sandbox? Were there any other directories where this was necessary?

Reply:
> though that is an assumption that you're making.

https://github.com/ollama/ollama/blob/5534f2cc6a3f29022998950472741d16e7a66b40/envconfig/config.go#L85-L100

- `OLLAMA_TMPDIR` should go into `/tmp`, together with `PrivateTmp=true`
- `OLLAMA_RUNNERS_DIR` not sure what that is

Reply:
> `OLLAMA_RUNNERS_DIR` not sure what that is

Basically, ollama compiles various runner binaries (cpu, cpu with simd, cuda, rocm) that are what actually run the ai models, then compresses and embeds these binaries into the main ollama binary at compile time. At runtime, they are decompressed and the gpu is tried first (cuda/rocm), falling back to a runner with fewer requirements (simd/cpu) upon failure.

I'm not sure what exactly `OLLAMA_RUNNERS_DIR` does, but one possibility is that this directory is where ollama extracts these runner binaries before executing them.

See [this comment](https://github.com/NixOS/nixpkgs/pull/323249#discussion_r1660362899) for a more detailed explanation.

Reply:
I would expect that they are then dumped into the wording directory, because even with hardening it still works for me. They are also probably not significant in size, when compared to the models.

Reply:
> Is there any other path that commonly needs to be writable?

It seems to me that `cfg.home` should also be added to `ReadWritePaths`.

Reply:
> I would expect that they are then dumped into the working directory

Actually they seem to land in /tmp, which is fine as well.

```
;evel=INFO source=payload.go:30 msg="extracting embedded files" dir=/tmp/ollama375027117/runners
```

> It seems to me that cfg.home should also be added to ReadWritePaths.

Possibly :thinking:

Reply:
If `cfg.home` isn't made writable, any custom home directory (eg `/home/user`) will cause errors and possibly crashes whenever ollama tries to write to any directory that isn't `${cfg.home}/models`.

That seems like a big problem to me, since the user can't make it writable with `writablePaths` anymore, and `systemd.services.ollama.serviceConfig.ReadWritePaths` has very low discoverability.

Reply:
Thanks for expanding on that, I was focusing on getting amdgpu to work in the meantime.

Added as prposed.